File: app/src/main/java/org/adaway/vpn/VpnService.java
/*
 * Derived from dns66:
 * Copyright (C) 2016-2019 Julian Andres Klode <jak@jak-linux.org>
 *
 * Derived from AdBuster:
 * Copyright (C) 2016 Daniel Brodie <dbrodie@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * Contributions shall also be provided under any later versions of the
 * GPL.
 */
package org.adaway.vpn;

import static android.Manifest.permission.POST_NOTIFICATIONS;
import static android.app.NotificationManager.IMPORTANCE_LOW;
import static android.app.PendingIntent.FLAG_IMMUTABLE;
import static android.content.Intent.FLAG_ACTIVITY_CLEAR_TASK;
import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;
import static android.content.pm.PackageManager.PERMISSION_GRANTED;
import static android.net.NetworkCapabilities.TRANSPORT_CELLULAR;
import static android.net.NetworkCapabilities.TRANSPORT_WIFI;
import static org.adaway.broadcast.Command.START;
import static org.adaway.broadcast.Command.STOP;
import static org.adaway.broadcast.CommandReceiver.SEND_COMMAND_ACTION;
import static org.adaway.helper.NotificationHelper.VPN_RESUME_SERVICE_NOTIFICATION_ID;
import static org.adaway.helper.NotificationHelper.VPN_RUNNING_SERVICE_NOTIFICATION_ID;
import static org.adaway.helper.NotificationHelper.VPN_SERVICE_NOTIFICATION_CHANNEL;
import static org.adaway.vpn.VpnService.NetworkType.CELLULAR;
import static org.adaway.vpn.VpnService.NetworkType.WIFI;
import static org.adaway.vpn.VpnStatus.RECONNECTING;
import static org.adaway.vpn.VpnStatus.RUNNING;
import static org.adaway.vpn.VpnStatus.STARTING;
import static org.adaway.vpn.VpnStatus.STOPPED;
import static org.adaway.vpn.VpnStatus.WAITING_FOR_NETWORK;
import static java.util.Objects.requireNonNull;

import android.app.Notification;
import android.app.PendingIntent;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.ConnectivityManager.NetworkCallback;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkRequest;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import org.adaway.R;
import org.adaway.broadcast.Command;
import org.adaway.broadcast.CommandReceiver;
import org.adaway.helper.PreferenceHelper;
import org.adaway.ui.home.HomeActivity;
import org.adaway.vpn.worker.VpnWorker;

import java.lang.ref.WeakReference;
import java.util.HashSet;
import java.util.Set;

import timber.log.Timber;

/**
 * This class is the VPN platform service implementation.
 * <p>
 * it is in charge of:
 * <ul>
 * <li>Accepting service commands,</li>
 * <li>Starting / stopping the {@link VpnWorker} thread,</li>
 * <li>Publishing notifications and intent about the VPN state,</li>
 * <li>Reacting to network connectivity changes.</li>
 * </ul>
 *
 * @author Bruce BUJON (bruce.bujon(at)gmail(dot)com)
 */
public class VpnService extends android.net.VpnService implements Handler.Callback {
    public static final String VPN_UPDATE_STATUS_INTENT = "org.jak_linux.dns66.VPN_UPDATE_STATUS";
    public static final String VPN_UPDATE_STATUS_EXTRA = "VPN_STATUS";
    /*
     * Notification intent related.
     */
    private static final int REQUEST_CODE_START = 43;
    private static final int REQUEST_CODE_PAUSE = 42;
    /*
     * Handler related.
     */
    private static final int VPN_STATUS_UPDATE_MESSAGE_TYPE = 0;

    private final MyHandler handler;
    private final NetworkTypeCallback wifiNetworkCallback;
    private final NetworkTypeCallback cellularNetworkCallback;
    private final Set<NetworkType> availableNetworkTypes;
    /**
     * The network the tunnel is currently bound to (whose DNS the {@link
     * org.adaway.vpn.dns.DnsServerMapper} resolved). When a network listed in
     * {@link #availableNetworkTypes} comes or goes but it is NOT this primary, there is
     * no need to rebuild the tunnel — the user keeps connectivity through the same
     * underlying network. This avoids tearing the tunnel down on every cellular
     * radio flicker on devices like Oppo / ColorOS that toggle the cellular transport
     * every few seconds while Wi-Fi is in use.
     */
    private NetworkType primaryNetwork;
    private final VpnWorker vpnWorker;

    /**
     * Constructor.
     */
    public VpnService() {
        this.handler = new MyHandler(this);
        this.wifiNetworkCallback = new NetworkTypeCallback(WIFI);
        this.cellularNetworkCallback = new NetworkTypeCallback(CELLULAR);
        this.availableNetworkTypes = new HashSet<>();
        this.vpnWorker = new VpnWorker(this);
    }

    /*
     * VPN Service.
     */

    @Override
    public void onCreate() {
        Timber.d("Creating VPN service…");
        registerNetworkCallback();
    }

    @Override
    public int onStartCommand(@Nullable Intent intent, int flags, int startId) {
        Timber.d("onStartCommand %s", intent == null ? "null intent" : intent);
        // Null intent means the system is resurrecting the service via START_STICKY.
        // https://developer.android.com/reference/android/app/Service#START_STICKY
        // If the user has explicitly disabled the VPN since we last ran, refuse the
        // resurrection — otherwise Android can silently bring the VPN back on its own
        // (see issues #4022 / #4234).
        if (intent == null) {
            boolean userEnabled = PreferenceHelper.getVpnServiceUserEnabled(this);
            if (!userEnabled) {
                Timber.i("Refusing sticky resurrection: user has disabled the VPN.");
                stopSelf(startId);
                return START_NOT_STICKY;
            }
        }
        Command command = intent == null ?
                START :
                Command.readFromIntent(intent);
        switch (command) {
            case START:
                startVpn();
                return START_STICKY;
            case STOP:
                stopVpn();
                return START_NOT_STICKY;
            default:
                Timber.w("Unknown command: %s", command);
                return START_NOT_STICKY;
        }
    }

    @Override
    public void onDestroy() {
        Timber.d("Destroying VPN service…");
        unregisterNetworkCallback();
        Timber.d("Destroyed VPN service.");
    }

    /*
     * Handler callback.
     */

    @Override
    public boolean handleMessage(@NonNull Message message) {
        if (message.what == VPN_STATUS_UPDATE_MESSAGE_TYPE) {
            updateVpnStatus(VpnStatus.fromCode(message.arg1));
        }
        return true;
    }

    /**
     * Notify a of the new VPN status.
     *
     * @param status The new VPN status.
     */
    public void notifyVpnStatus(VpnStatus status) {
        Message statusMessage = this.handler.obtainMessage(VPN_STATUS_UPDATE_MESSAGE_TYPE, status.toCode(), 0);
        this.handler.sendMessage(statusMessage);
    }

    private void startVpn() {
        Timber.d("Starting VPN service…");
        PreferenceHelper.setVpnServiceStatus(this, RUNNING);
        updateVpnStatus(STARTING);
        // This path is reached only via an explicit START intent (user click, notif
        // action, autostart, sticky resurrection that survived the user-intent gate).
        // The throttler is meant to dampen reconnection storms, NOT to delay user
        // actions — reset it so the tunnel comes up immediately.
        this.vpnWorker.resetThrottle();
        this.vpnWorker.start();
        Timber.i("VPN service started.");
    }

    private void stopVpn() {
        Timber.d("Stopping VPN service…");
        PreferenceHelper.setVpnServiceStatus(this, STOPPED);
        this.vpnWorker.stop();
        stopForeground(true);
        stopSelf();
        updateVpnStatus(STOPPED);
        Timber.i("VPN service stopped.");
    }

    private void waitForNetVpn() {
        this.vpnWorker.stop();
        updateVpnStatus(WAITING_FOR_NETWORK);
    }

    private void reconnect() {
        updateVpnStatus(RECONNECTING);
        this.vpnWorker.start();
    }

    private void updateVpnStatus(VpnStatus status) {
        Notification notification = getNotification(status);
        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(this);
        switch (status) {
            case STARTING:
            case RUNNING:
                notificationManager.cancel(VPN_RESUME_SERVICE_NOTIFICATION_ID);
                startForeground(VPN_RUNNING_SERVICE_NOTIFICATION_ID, notification);
                break;
            default:
                if (checkSelfPermission(POST_NOTIFICATIONS) == PERMISSION_GRANTED) {
                    notificationManager.notify(VPN_RESUME_SERVICE_NOTIFICATION_ID, notification);
                }
        }

        // TODO BUG - Nobody is listening to this intent
        // TODO BUG - VpnModel can lister to it to update the MainActivity according its current state
        Intent intent = new Intent(VPN_UPDATE_STATUS_INTENT);
        intent.putExtra(VPN_UPDATE_STATUS_EXTRA, status);
        LocalBroadcastManager.getInstance(this).sendBroadcast(intent);
    }

    private Notification getNotification(VpnStatus status) {
        String title = getString(R.string.vpn_notification_title, getString(status.getTextResource()));

        Intent intent = new Intent(getApplicationContext(), HomeActivity.class);
        intent.setFlags(FLAG_ACTIVITY_NEW_TASK | FLAG_ACTIVITY_CLEAR_TASK);
        PendingIntent contentIntent = PendingIntent.getActivity(getApplicationContext(), 0, intent, FLAG_IMMUTABLE);

        NotificationCompat.Builder builder = new NotificationCompat.Builder(this, VPN_SERVICE_NOTIFICATION_CHANNEL)
                .setPriority(IMPORTANCE_LOW)
                .setContentIntent(contentIntent)
                .setSmallIcon(R.drawable.logo)
                .setColorized(true)
                .setColor(getColor(R.color.notification))
                .setContentTitle(title);
        switch (status) {
            case RUNNING:
                Intent stopIntent = new Intent(this, CommandReceiver.class)
                        .setAction(SEND_COMMAND_ACTION);
                STOP.appendToIntent(stopIntent);
                PendingIntent stopActionIntent = PendingIntent.getBroadcast(this, REQUEST_CODE_PAUSE, stopIntent, FLAG_IMMUTABLE);
                builder.addAction(
                        R.drawable.ic_pause_24dp,
                        getString(R.string.vpn_notification_action_pause),
                        stopActionIntent
                ).setOngoing(true);
                break;
            case STOPPED:
                Intent startIntent = new Intent(this, CommandReceiver.class)
                        .setAction(SEND_COMMAND_ACTION);
                START.appendToIntent(startIntent);
                PendingIntent startActionIntent = PendingIntent.getBroadcast(this, REQUEST_CODE_START, startIntent, FLAG_IMMUTABLE);
                builder.addAction(
                        0,
                        getString(R.string.vpn_notification_action_resume),
                        startActionIntent
                );
                break;
        }
        return builder.build();
    }

    private void registerNetworkCallback() {
        ConnectivityManager connectivityManager = (ConnectivityManager) getSystemService(CONNECTIVITY_SERVICE);
        NetworkRequest wifiNetworkRequest = new NetworkRequest.Builder()
                .addTransportType(TRANSPORT_WIFI)
                .build();
        NetworkRequest cellularNetworkRequest = new NetworkRequest.Builder()
                .addTransportType(TRANSPORT_CELLULAR)
                .build();
        initializeNetworkTypes(connectivityManager);
        connectivityManager.registerNetworkCallback(wifiNetworkRequest, this.wifiNetworkCallback, this.handler);
        connectivityManager.registerNetworkCallback(cellularNetworkRequest, this.cellularNetworkCallback, this.handler);

    }

    private void unregisterNetworkCallback() {
        ConnectivityManager connectivityManager = (ConnectivityManager) getSystemService(CONNECTIVITY_SERVICE);
        connectivityManager.unregisterNetworkCallback(this.wifiNetworkCallback);
        connectivityManager.unregisterNetworkCallback(this.cellularNetworkCallback);
    }

    private void initializeNetworkTypes(ConnectivityManager connectivityManager) {
        this.availableNetworkTypes.clear();
        this.primaryNetwork = null;
        Network activeNetwork = connectivityManager.getActiveNetwork();
        if (activeNetwork != null) {
            NetworkCapabilities networkCapabilities = connectivityManager.getNetworkCapabilities(activeNetwork);
            if (networkCapabilities != null) {
                if (networkCapabilities.hasTransport(TRANSPORT_WIFI)) {
                    this.availableNetworkTypes.add(WIFI);
                    this.primaryNetwork = WIFI;
                }
                if (networkCapabilities.hasTransport(TRANSPORT_CELLULAR)) {
                    this.availableNetworkTypes.add(CELLULAR);
                    if (this.primaryNetwork == null) {
                        this.primaryNetwork = CELLULAR;
                    }
                }
            }
        }
        Timber.d("Initial network types: %s, primary=%s.", this.availableNetworkTypes, this.primaryNetwork);
    }

    private void addNetworkType(NetworkType type) {
        boolean noNetwork = this.availableNetworkTypes.isEmpty();
        this.availableNetworkTypes.add(type);
        if (noNetwork) {
            this.primaryNetwork = type;
            Timber.d("Reconnecting VPN on network %s.", type);
            reconnect();
        } else {
            // Adding a secondary network does not require a tunnel rebuild — the
            // tunnel is still bound to the primary network whose DNS we resolved.
            Timber.d("Secondary network %s available, keeping tunnel on %s.", type, this.primaryNetwork);
        }
    }

    private void removeNetworkType(NetworkType type) {
        this.availableNetworkTypes.remove(type);
        if (this.availableNetworkTypes.isEmpty()) {
            this.primaryNetwork = null;
            Timber.d("Waiting for network…");
            waitForNetVpn();
        } else if (type == this.primaryNetwork) {
            // The network the tunnel was bound to is gone but a fallback is still
            // available; switch over so the DNS server mapper can re-resolve.
            this.primaryNetwork = this.availableNetworkTypes.iterator().next();
            Timber.d("Primary network %s lost, reconnecting on %s.", type, this.primaryNetwork);
            reconnect();
        } else {
            // A secondary network went away while the primary is still up — keep the
            // tunnel as-is. Without this guard the VPN was being torn down and rebuilt
            // every time the cellular radio flickered on Oppo / ColorOS power-saving
            // devices, making the status-bar VPN icon blink continuously.
            Timber.d("Secondary network %s lost, keeping tunnel on %s.", type, this.primaryNetwork);
        }
    }

    /**
     * This class receives network change events to monitor network type available.
     *
     * @author Bruce BUJON (bruce.bujon(at)gmail(dot)com)
     * @see <a href="https://developer.android.com/training/basics/network-ops/reading-network-state#listening-events">Android Developer Documentation</a>
     */
    private class NetworkTypeCallback extends NetworkCallback {
        private final NetworkType monitoredType;

        NetworkTypeCallback(NetworkType monitoredType) {
            this.monitoredType = monitoredType;
        }

        @Override
        public void onAvailable(@NonNull Network network) {
            Timber.d("On available %s", this.monitoredType);
            addNetworkType(this.monitoredType);
        }

        @Override
        public void onLost(@NonNull Network network) {
            Timber.d("On lost %s", this.monitoredType);
            removeNetworkType(this.monitoredType);
        }
    }

    enum NetworkType {
        CELLULAR,
        WIFI,
    }

    /* The handler may only keep a weak reference around, otherwise it leaks */
    private static class MyHandler extends Handler {

        private final WeakReference<Callback> callback;

        MyHandler(Callback callback) {
            super(requireNonNull(Looper.myLooper()));
            this.callback = new WeakReference<>(callback);
        }

        @Override
        public void handleMessage(@NonNull Message msg) {
            Callback callback = this.callback.get();
            if (callback != null) {
                callback.handleMessage(msg);
            }
            super.handleMessage(msg);
        }
    }
}

File: app/src/main/java/org/adaway/vpn/VpnService.kt
package org.adaway.vpn

import android.Manifest.permission.POST_NOTIFICATIONS
import android.app.Notification
import android.app.PendingIntent
import android.app.PendingIntent.FLAG_IMMUTABLE
import android.app.Service.STOP_FOREGROUND_REMOVE
import android.content.Intent
import android.content.Intent.FLAG_ACTIVITY_CLEAR_TASK
import android.content.Intent.FLAG_ACTIVITY_NEW_TASK
import android.content.pm.PackageManager.PERMISSION_GRANTED
import android.content.pm.ServiceInfo
import android.net.ConnectivityManager
import android.net.ConnectivityManager.NetworkCallback
import android.net.Network
import android.net.NetworkCapabilities.TRANSPORT_CELLULAR
import android.net.NetworkCapabilities.TRANSPORT_WIFI
import android.net.NetworkRequest
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.Message
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import org.adaway.R
import org.adaway.broadcast.Command
import org.adaway.broadcast.Command.START
import org.adaway.broadcast.Command.STOP
import org.adaway.broadcast.CommandReceiver
import org.adaway.broadcast.CommandReceiver.Companion.SEND_COMMAND_ACTION
import org.adaway.helper.NotificationHelper.VPN_RESUME_SERVICE_NOTIFICATION_ID
import org.adaway.helper.NotificationHelper.VPN_RUNNING_SERVICE_NOTIFICATION_ID
import org.adaway.helper.NotificationHelper.VPN_SERVICE_NOTIFICATION_CHANNEL
import org.adaway.helper.PreferenceHelper
import org.adaway.ui.home.HomeActivity
import org.adaway.vpn.VpnStatus.RECONNECTING
import org.adaway.vpn.VpnStatus.RUNNING
import org.adaway.vpn.VpnStatus.STARTING
import org.adaway.vpn.VpnStatus.STOPPED
import org.adaway.vpn.VpnStatus.WAITING_FOR_NETWORK
import org.adaway.vpn.worker.VpnWorker
import timber.log.Timber
import java.lang.ref.WeakReference

class VpnService : android.net.VpnService(), Handler.Callback {
    private val handler = MyHandler(this)
    private val wifiNetworkCallback = NetworkTypeCallback(NetworkType.WIFI)
    private val cellularNetworkCallback = NetworkTypeCallback(NetworkType.CELLULAR)
    private val availableNetworkTypes = mutableSetOf<NetworkType>()
    private val vpnWorker = VpnWorker(this)

    override fun onCreate() {
        Timber.d("Creating VPN service...")
        registerNetworkCallback()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Timber.d("onStartCommand %s", intent ?: "null intent")
        return when (val command = intent?.let(Command::readFromIntent) ?: START) {
            START -> {
                startVpn()
                START_STICKY
            }

            STOP -> {
                stopVpn()
                START_NOT_STICKY
            }

            Command.UNKNOWN -> {
                Timber.w("Unknown command: %s", command)
                START_NOT_STICKY
            }
        }
    }

    override fun onDestroy() {
        Timber.d("Destroying VPN service...")
        unregisterNetworkCallback()
        Timber.d("Destroyed VPN service.")
    }

    override fun handleMessage(message: Message): Boolean {
        if (message.what == VPN_STATUS_UPDATE_MESSAGE_TYPE) {
            updateVpnStatus(VpnStatus.fromCode(message.arg1))
        }
        return true
    }

    fun notifyVpnStatus(status: VpnStatus) {
        val statusMessage = handler.obtainMessage(VPN_STATUS_UPDATE_MESSAGE_TYPE, status.toCode(), 0)
        handler.sendMessage(statusMessage)
    }

    private fun startVpn() {
        Timber.d("Starting VPN service...")
        PreferenceHelper.setVpnServiceStatus(this, RUNNING)
        updateVpnStatus(STARTING)
        vpnWorker.start()
        Timber.i("VPN service started.")
    }

    private fun stopVpn() {
        Timber.d("Stopping VPN service...")
        PreferenceHelper.setVpnServiceStatus(this, STOPPED)
        vpnWorker.stop()
        stopForeground(STOP_FOREGROUND_REMOVE)
        stopSelf()
        updateVpnStatus(STOPPED)
        Timber.i("VPN service stopped.")
    }

    private fun waitForNetVpn() {
        vpnWorker.stop()
        updateVpnStatus(WAITING_FOR_NETWORK)
    }

    private fun reconnect() {
        updateVpnStatus(RECONNECTING)
        vpnWorker.start()
    }

    private fun updateVpnStatus(status: VpnStatus) {
        val notification = getNotification(status)
        val notificationManager = NotificationManagerCompat.from(this)
        when (status) {
            STARTING, RUNNING -> {
                notificationManager.cancel(VPN_RESUME_SERVICE_NOTIFICATION_ID)
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                    startForeground(
                        VPN_RUNNING_SERVICE_NOTIFICATION_ID,
                        notification,
                        ServiceInfo.FOREGROUND_SERVICE_TYPE_SPECIAL_USE
                    )
                } else {
                    startForeground(VPN_RUNNING_SERVICE_NOTIFICATION_ID, notification)
                }
            }

            else -> {
                if (checkSelfPermission(POST_NOTIFICATIONS) == PERMISSION_GRANTED) {
                    notificationManager.notify(VPN_RESUME_SERVICE_NOTIFICATION_ID, notification)
                }
            }
        }
        VpnStatusRepository.update(status)
    }

    private fun getNotification(status: VpnStatus): Notification {
        val title = getString(R.string.vpn_notification_title, getString(status.textResource))
        val intent = Intent(applicationContext, HomeActivity::class.java)
            .setFlags(FLAG_ACTIVITY_NEW_TASK or FLAG_ACTIVITY_CLEAR_TASK)
        val contentIntent = PendingIntent.getActivity(applicationContext, 0, intent, FLAG_IMMUTABLE)

        val builder = NotificationCompat.Builder(this, VPN_SERVICE_NOTIFICATION_CHANNEL)
            .setPriority(android.app.NotificationManager.IMPORTANCE_LOW)
            .setContentIntent(contentIntent)
            .setSmallIcon(R.drawable.logo)
            .setColor(getColor(R.color.notification))
            .setContentTitle(title)

        if (Build.VERSION.SDK_INT >= 36) {
            val notificationManager = getSystemService(android.app.NotificationManager::class.java)
            if (notificationManager != null && notificationManager.canPostPromotedNotifications()) {
                builder.extras.putBoolean("android.requestPromotedOngoing", true)
                builder.setColorized(false)
            } else {
                builder.setColorized(true)
            }
        } else {
            builder.setColorized(true)
        }

        when (status) {
            RUNNING -> {
                val stopIntent = Intent(this, CommandReceiver::class.java)
                    .setAction(SEND_COMMAND_ACTION)
                STOP.appendToIntent(stopIntent)
                val stopActionIntent = PendingIntent.getBroadcast(this, REQUEST_CODE_PAUSE, stopIntent, FLAG_IMMUTABLE)
                builder.addAction(
                    R.drawable.ic_pause_24dp,
                    getString(R.string.vpn_notification_action_pause),
                    stopActionIntent
                ).setOngoing(true)
            }

            STOPPED -> {
                val startIntent = Intent(this, CommandReceiver::class.java)
                    .setAction(SEND_COMMAND_ACTION)
                START.appendToIntent(startIntent)
                val startActionIntent = PendingIntent.getBroadcast(this, REQUEST_CODE_START, startIntent, FLAG_IMMUTABLE)
                builder.addAction(
                    0,
                    getString(R.string.vpn_notification_action_resume),
                    startActionIntent
                )
            }

            else -> Unit
        }
        return builder.build()
    }

    private fun registerNetworkCallback() {
        val connectivityManager = getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
        val wifiNetworkRequest = NetworkRequest.Builder()
            .addTransportType(TRANSPORT_WIFI)
            .build()
        val cellularNetworkRequest = NetworkRequest.Builder()
            .addTransportType(TRANSPORT_CELLULAR)
            .build()
        initializeNetworkTypes(connectivityManager)
        connectivityManager.registerNetworkCallback(wifiNetworkRequest, wifiNetworkCallback, handler)
        connectivityManager.registerNetworkCallback(cellularNetworkRequest, cellularNetworkCallback, handler)
    }

    private fun unregisterNetworkCallback() {
        val connectivityManager = getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
        connectivityManager.unregisterNetworkCallback(wifiNetworkCallback)
        connectivityManager.unregisterNetworkCallback(cellularNetworkCallback)
    }

    private fun initializeNetworkTypes(connectivityManager: ConnectivityManager) {
        availableNetworkTypes.clear()
        connectivityManager.activeNetwork?.let { activeNetwork ->
            connectivityManager.getNetworkCapabilities(activeNetwork)?.let { networkCapabilities ->
                if (networkCapabilities.hasTransport(TRANSPORT_WIFI)) {
                    availableNetworkTypes += NetworkType.WIFI
                }
                if (networkCapabilities.hasTransport(TRANSPORT_CELLULAR)) {
                    availableNetworkTypes += NetworkType.CELLULAR
                }
            }
        }
        Timber.d("Initial network types: %s ", availableNetworkTypes)
    }

    private fun addNetworkType(type: NetworkType) {
        val noNetwork = availableNetworkTypes.isEmpty()
        availableNetworkTypes += type
        if (noNetwork) {
            Timber.d("Reconnecting VPN on network %s.", type)
            reconnect()
        }
    }

    private fun removeNetworkType(type: NetworkType) {
        availableNetworkTypes -= type
        if (availableNetworkTypes.isEmpty()) {
            Timber.d("Waiting for network...")
            waitForNetVpn()
        } else {
            reconnect()
        }
    }

    private inner class NetworkTypeCallback(private val monitoredType: NetworkType) : NetworkCallback() {
        override fun onAvailable(network: Network) {
            Timber.d("On available %s", monitoredType)
            addNetworkType(monitoredType)
        }

        override fun onLost(network: Network) {
            Timber.d("On lost %s", monitoredType)
            removeNetworkType(monitoredType)
        }
    }

    private enum class NetworkType {
        CELLULAR,
        WIFI,
    }

    private class MyHandler(callback: Handler.Callback) : Handler(Looper.getMainLooper()) {
        private val callback = WeakReference(callback)

        override fun handleMessage(msg: Message) {
            callback.get()?.handleMessage(msg)
            super.handleMessage(msg)
        }
    }

    companion object {
        private const val REQUEST_CODE_START = 43
        private const val REQUEST_CODE_PAUSE = 42
        private const val VPN_STATUS_UPDATE_MESSAGE_TYPE = 0
    }
}
